### ISSUE TITLE
refFilter*Regex options "don't work"(tm) 

### ISSUE THREAD

**Issue by pfalcon** [git-multimail/git-multimail#120]
To add to confusing reports I submitted (#118, #119) here's another one. (And I do understand they're confusing, but working in devops, with bunch of issues and projects in works, I can't get to investigate them in details for days/weeks, so figure I'd submit them as our users face them - I hope that provides an interesting perspective too: users just do everything per docs, and it just doesn't work :-D).

So, we use gitolite (as in the issues above), and have to projects:

1. Uses:

~~~~
    config multimailhook.refFilterInclusionRegex = "linux-linaro-lsk-v\d\.\d+($|-rt|-android)$"
~~~~

Notification mails get sent out even for branches not matching the regex.

2. Uses:

~~~~
     config multimailhook.refFilterExclusionRegex = "/trunk$"
~~~~

Yes, you guessed - notification mails are still sent for branch named "trunk" (it's SVN-imported git repo).

**Comment by moy** [comment]
>  I hope that provides an interesting perspective too

It sure does. Continue sending user feedback (or better, pull-requests ;-) ).

Can you reproduce the issue outside gitolite? Try to reproduce this on a toy repository and give a reproduction recipe. Maybe @newren can help as he's the author of this feature.

**Comment by pfalcon** [comment]
Now, what could get wrong?

1. We use wrong regex format; for example, you do .match() on regexes, whereas my examples above assume .search(). But why do you do this to us? Mere humans don't usually want to filter "references" (sic!, even if you name an option like that, they won't believe you want to force them to use full reference path every time), they usually want to filter just branches, and don't care about ambiguities, because in handful branches they have, there can't be any. And folks who need full references, them would use anchored regex, `^refs/...`.

2. Quoting issues passing values thru gitolite (I did follow this path and found it clean few weeks ago, but well, it was few weeks ago).

3. We don't get distinction between refFilterDoSendRegex and refFilterInclusionRegex, etc.

4. Something else.

**Comment by pfalcon** [comment]
> It sure does. Continue sending user feedback (or better, pull-requests ;-) ).

Hope to get to it, just between getting a summer vacation and upgrading few gerrit installs in use by hundreds of users, it's not easy to find some time to concentrate on git-multimail ;-).  

> Can you reproduce the issue outside gitolite? Try to reproduce this on a toy repository and give a reproduction recipe. Maybe @newren can help as he's the author of this feature.

Not yet. Hoped to look into it this week, but hitting a Friday evening, figured I'd at least dump what I know so far, and do that in official project tracker rather than our internal, in the hope that it'll be more useful than a noise. I do intend to follow myself each of tickets I submitted, but hints and guidance is appreciated.

Thanks.

**Comment by moy** [comment]
> upgrading few gerrit installs in use by hundreds of users

Note that you are using unreleased features. Especially the refFilter* stuff is new and may still change depending on how users like the current implementation. Using it in production may not be the best idea (but I do appreciate user feedback especially on that feature).

> you do .match() on regexes, whereas my examples above assume .search().

No, we do `.search()`: https://github.com/git-multimail/git-multimail/blob/master/git-multimail/git_multimail.py#L3271

I'm adding a test in 2d87f3133cac61783b588a6f0107b5bef93f8c38 to make sure we don't break it.

> Mere humans don't usually want to filter "references"

Mere humans do not usually want to receive notifications for `refs/notes/` or `refs/whatever-namespace-your-local-tool-use`, and want to filter them out. So it's important to match against the whole refname to allow that. Internally, we do use this to ignore the gerrit-specific refs.

> Quoting issues passing values thru gitolite

I'm not enough of a gitolite expert to say. But you should check the actual Git config files on the gitolite server if you have access to it.

> We don't get distinction between refFilterDoSendRegex and refFilterInclusionRegex, etc.

This should make no difference for refChange emails. For commit emails, refFilterInclusionRegex will still send a commit email when the commit reaches a ref that is not ignored. refFilterDoSendRegex will not send an email at all. Well, modulo bugs ;-).

**Comment by moy** [comment]
Updated the doc in 9640f2ecd405e5bf5f7449fc55fd9bc231bc649d too.

**Comment by moy** [comment]
Just tried on my gitolite install:

```
repo @all
    # Not strictly needed as git_multimail.py will chose gitolite if
    # $GL_USER is set.
    config multimailhook.environment = gitolite
    config multimailhook.mailinglist = <my email>
    config multimailhook.refFilterExclusionRegex = "/branch$"
    config multimailhook.from = ""
```

Pushing to branch `branch` sends nothing. Pushing to another branch does send. So, "it works"(tm) ...

**Comment by pfalcon** [comment]
Thanks for all the replies, much appreciated. I'll get to them next week (can't promise Monday). One of the issues is that repos mentioned above aren't toy at all - one is Linux Kernel fork, another GCC fork. The ticket title is intended to be humorous, I certainly don't mean that these options don't work at all - only that in real-world setting (and on "more complex" side of spectrum), they don't behave as users (us) expect. And given that issues happen on huge/complex repos, I don't expect easy investigation, but then it means need to carefully write down all the facts/progress, which this ticket starts.

**Comment by moy** [comment]
Sorry if my last comment appeared offensive, I should have added a smiley after the (tm) ;-).

But if I can't reproduce, you'll have to be the one doing the work of narrowing down the issue :-(.

**Comment by pfalcon** [comment]
> But if I can't reproduce, you'll have to be the one doing the work of narrowing down the issue :-(.

Sounds good for me! ;-)

**Comment by pfalcon** [comment]
> Note that you are using unreleased features. Especially the refFilter* stuff is new and may still change depending on how users like the current implementation. Using it in production may not be the best idea (but I do appreciate user feedback especially on that feature).

Yes, and when I was approached to install git-multimail, I conservatively asked "why can't we use standard post-receive-email script?" Support for filtering branches and better detection which commit belongs to which branch were quoted among the reasons. So, we're happy to test-drive it, report any issue, and dig in them, as time permits.

**Comment by pfalcon** [comment]
For reference, repos quoted above are public:

1. Linaro LSK kernel:
https://git.linaro.org/kernel/linux-linaro-stable.git
Notifications archive (may be on and off, we disable posting to it if see big issues): https://lists.linaro.org/pipermail/lsk-commits/

2. Linaro GCC mirror:
https://git.linaro.org/toolchain/gcc.git
Notifications archive: https://lists.linaro.org/pipermail/tcwg-commits/


(Just to clarify, I don't expect anyone to drop everything and go do something with the above - it's only to collect all related info in one place, for future reference.)

**Comment by moy** [comment]
> why can't we use standard post-receive-email script?

The other reason is: git-multimail obsoleted it 2 years ago: 1d25e77 (post-receive-email: deprecate script in favor of git-multimail, 2013-07-14). So, you would need to `s/standard/old/` above.

In any case, I'm happy to have real-life testers.

**Comment by pfalcon** [comment]
Ok, so here's funny thing. Initially (from 2015-07-08), we ran git-multimail 1.1.1 (from the git tag). Then, before starting to investigating reported issues in this and other tickets, and seeing there was bunch of commits to git_multimail master, I reviewed every commit from 1.1.1 to 1560964e45ad890caa7 (HEAD at that time), to see if: a) there may be fixes for `refFilter*`; b) there may be changes which would affect us adversely. I saw none `(*)`, so on 2015-08-21, I upgraded to the version above.

Then, couple days later, an LSK maintainer let me know that "branch filtering works as expected since upgrade". "Ok", I thought "you guys don't want for gazillion of test commits to land in your repo. Well, we have another one." So, today I looked at recent gcc mails, and - issue indeed gone!

So I wasn't called a liar, here's the latest email which was sent for "trunk" branch (again, the repo is an git-svn import): https://lists.linaro.org/pipermail/tcwg-commits/2015-August/001785.html . If you look at the mail archive, there're whole bunch of mails for trunk:  https://lists.linaro.org/pipermail/tcwg-commits/2015-August/date.html before that, and none since.




`(*)` Except for supporting multiple values for refFilter*, but who would pay attention to that, right?
